### PR TITLE
Write DelayedArray to file on the R side.

### LINKED ISSUE #32: Skip DelayedArray matrices in SCE2AnnData()
Some **DelayedArray** matrices cannot be automatically converted to `dgCMatrix` and fail conversion to standard matrices due to memory limitations. See https://support.bioconductor.org/p/p134226/#p134234 for original issue.

We should detect these automatically and either skip them (with a warning) or manage the conversion (if possible).

### PR DESCRIPTION
**Pull request by LTLA**
Closes #32. 

```r
library(zellkonverter)
library(scRNAseq)
sce <- ZeiselBrainData()
counts(sce) <- DelayedArray(counts(sce))
 
temp <- tempfile(fileext = ".h5ad")
writeH5AD(sce, temp)
```

Some points:

- Need to test that it works with both DA in the first and/or later assays.
- Need to test that the result is still readable by `readH5AD`.

Known to-dos:

- The code does not respect `is_sparse()` being `TRUE`. A little bit of work is required to achieve a one-pass writer that does not rely on knowing the total number of non-zero entries.

### CODE REVIEW

File: R/write.R [hunk 1]
```diff
@@ -50,13 +50,45 @@
...
         fun = .H5ADwriter,
         sce = sce, file = file, X_name = X_name, skip_assays = skip_assays
     )
+
+    # Going back out and replacing each of them.
+    if (any(is_da)) {
+        paths <- file.path("layers", assayNames(sce)[is_da])
+        if (is_da[1]) {
+            paths[1] <- "X"
+        }
```
Comment (lazappi) [inline R/write.R:78]: Is this redundant with the check within the loop? Just wondering if we could cut one of them?

Comment (LTLA) [reply R/write.R:78]: Yeah, probably. Must have started writing it outside the loop and realized I needed to move it inside and forgot to delete the former. I'd just move the `file.path` call inside the loop as well, no point splitting the logic.

**Comment by LTLA** [comment]
Latest commit solves the to-do above, but needs a lot of testing.

Briefly, we set up extensible HDF5 datasets and then iterate across row-wise chunks of the sparse DA. In each chunk, we extract the non-zero values, convert them to compressed sparse row format, and append them to the existing HDF5 dataset via `h5set_extent` (to make the existing dataset larger) and `h5writeDataset` to write the new values in the expanded space. This approach respects the memory constraints of block processing, by ensuring that we do not load all non-zero values into memory; while retaining a single-pass approach for rapid processing, by avoiding the need to know the total number of non-zeroes.

The code here makes a number of assumptions:

- I used `values`, `indptrs` and `index` for the values, column indices and pointers, respectively. Can't remember what H5AD actually calls them. Writing this now, I suspect I mixed up `index` and `indptrs`.
- Column indices are assumed to be zero-based.
- Stored the pointers as 64-bit unsigned integers as the number of non-zeros can exceed ~3 billion for large datasets.

**Comment by lazappi** [comment]
Added a couple of tests and fixed how the loop for rewriting matrices handled paths.

**Comment by LTLA** [comment]
Hit a roadblock in the form of grimbough/rhdf5#79; the AnnData reader doesn't like the fixed-width byte strings that **rhdf5** emits.

For the time being, I suggest we just add a clause to the initial check for DAs where we do not skip a DA if it `is_sparse()` is TRUE. This implies that it will be realized into memory in `SCE2AnnData` - hopefully this is tolerated on the user machine. We can switch back to the current block processing behavior once the **rhdf5** issue is resolved.

**Comment by LTLA** [comment]
@lazappi were you going to work on this?

**Comment by lazappi** [comment]
Sorry I've lost track of it a bit. Is it just the `is_sparse()` check that we need to add?

**Comment by LTLA** [comment]
Yes, I think line 60 in my PR should only skip the assay if it's a DA and `!is_sparse()`. Sparse DAs should be coerced to in-memory `dgCMatrix` objects for the time being.

**Comment by lazappi** [comment]
Ok, I've added that check. The code for writing sparse DelayedArrays can't be reached by anything now which made the test coverage drop but that's fine and better to have it there for when the **rhdf5** issue is fixed. Anything else to add?

**Comment by LTLA** [comment]
Think that's it, just slap on some `# nocov start` and `end`s around it for the time being, I guess.

**Pull request merged**